File: benchmarks/values/struct_nbody.py
import time
import math

t0 = time.perf_counter()

PI = 3.141592653589793
SOLAR_MASS = 4.0 * PI * PI
DAYS_PER_YEAR = 365.24

# Bodies: [x, y, z, vx, vy, vz, mass]
bodies = [
    # Sun
    [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, SOLAR_MASS],
    # Jupiter
    [4.84143144246472090e+00, -1.16032004402742839e+00, -1.03622044471123109e-01,
     1.66007664274403694e-03 * DAYS_PER_YEAR,
     7.69901118419740425e-03 * DAYS_PER_YEAR,
     -6.90460016972063023e-05 * DAYS_PER_YEAR,
     9.54791938424326609e-04 * SOLAR_MASS],
    # Saturn
    [8.34336671824457987e+00, 4.12479856412430479e+00, -4.03523417114321381e-01,
     -2.76742510726862411e-03 * DAYS_PER_YEAR,
     4.99852801234917238e-03 * DAYS_PER_YEAR,
     2.30417297573763929e-05 * DAYS_PER_YEAR,
     2.85885980666130812e-04 * SOLAR_MASS],
    # Uranus
    [1.28943695621391310e+01, -1.51111514016986312e+01, -2.23307578892655734e-01,
     2.96460137564761618e-03 * DAYS_PER_YEAR,
     2.37847173959480950e-03 * DAYS_PER_YEAR,
     -2.96589568540237556e-05 * DAYS_PER_YEAR,
     4.36624404335156298e-05 * SOLAR_MASS],
    # Neptune
    [1.53796971148509165e+01, -2.59193146099879641e+01, 1.79258772950371181e-01,
     2.68067772490389322e-03 * DAYS_PER_YEAR,
     1.62824170038242295e-03 * DAYS_PER_YEAR,
     -9.51592254519715870e-05 * DAYS_PER_YEAR,
     5.15138902046611451e-05 * SOLAR_MASS],
]

def offset_momentum(bodies):
    px = py = pz = 0.0
    for b in bodies:
        px += b[3] * b[6]
        py += b[4] * b[6]
        pz += b[5] * b[6]
    bodies[0][3] = -px / SOLAR_MASS
    bodies[0][4] = -py / SOLAR_MASS
    bodies[0][5] = -pz / SOLAR_MASS

def advance(bodies, dt):
    n = len(bodies)
    for i in range(n):
        bi = bodies[i]
        for j in range(i + 1, n):
            bj = bodies[j]
            dx = bi[0] - bj[0]
            dy = bi[1] - bj[1]
            dz = bi[2] - bj[2]
            dist = math.sqrt(dx * dx + dy * dy + dz * dz)
            mag = dt / (dist * dist * dist)
            bi[3] -= dx * bj[6] * mag
            bi[4] -= dy * bj[6] * mag
            bi[5] -= dz * bj[6] * mag
            bj[3] += dx * bi[6] * mag
            bj[4] += dy * bi[6] * mag
            bj[5] += dz * bi[6] * mag
    for b in bodies:
        b[0] += dt * b[3]
        b[1] += dt * b[4]
        b[2] += dt * b[5]

def energy(bodies):
    e = 0.0
    n = len(bodies)
    for i in range(n):
        bi = bodies[i]
        e += 0.5 * bi[6] * (bi[3] * bi[3] + bi[4] * bi[4] + bi[5] * bi[5])
        for j in range(i + 1, n):
            bj = bodies[j]
            dx = bi[0] - bj[0]
            dy = bi[1] - bj[1]
            dz = bi[2] - bj[2]
            dist = math.sqrt(dx * dx + dy * dy + dz * dz)
            e -= bi[6] * bj[6] / dist
    return e

offset_momentum(bodies)

steps = 500_000
dt = 0.01
for _ in range(steps):
    advance(bodies, dt)

e = energy(bodies)

t1 = time.perf_counter()
print(f"energy: {e:.9f}")
print(f"elapsed: {t1 - t0:.3f}s")

File: benchmarks/values/struct_nbody.go
package main

import (
	"fmt"
	"math"
	"time"
)

const (
	pi          = 3.141592653589793
	solarMass   = 4.0 * pi * pi
	daysPerYear = 365.24
	nBodies     = 5
)

type Body struct {
	x, y, z, vx, vy, vz, mass float64
}

func advance(bodies []Body, dt float64) {
	for i := 0; i < nBodies; i++ {
		for j := i + 1; j < nBodies; j++ {
			dx := bodies[i].x - bodies[j].x
			dy := bodies[i].y - bodies[j].y
			dz := bodies[i].z - bodies[j].z
			dist := math.Sqrt(dx*dx + dy*dy + dz*dz)
			mag := dt / (dist * dist * dist)
			bodies[i].vx -= dx * bodies[j].mass * mag
			bodies[i].vy -= dy * bodies[j].mass * mag
			bodies[i].vz -= dz * bodies[j].mass * mag
			bodies[j].vx += dx * bodies[i].mass * mag
			bodies[j].vy += dy * bodies[i].mass * mag
			bodies[j].vz += dz * bodies[i].mass * mag
		}
	}
	for i := 0; i < nBodies; i++ {
		bodies[i].x += dt * bodies[i].vx
		bodies[i].y += dt * bodies[i].vy
		bodies[i].z += dt * bodies[i].vz
	}
}

func energy(bodies []Body) float64 {
	e := 0.0
	for i := 0; i < nBodies; i++ {
		e += 0.5 * bodies[i].mass *
			(bodies[i].vx*bodies[i].vx +
				bodies[i].vy*bodies[i].vy +
				bodies[i].vz*bodies[i].vz)
		for j := i + 1; j < nBodies; j++ {
			dx := bodies[i].x - bodies[j].x
			dy := bodies[i].y - bodies[j].y
			dz := bodies[i].z - bodies[j].z
			dist := math.Sqrt(dx*dx + dy*dy + dz*dz)
			e -= bodies[i].mass * bodies[j].mass / dist
		}
	}
	return e
}

func offsetMomentum(bodies []Body) {
	px, py, pz := 0.0, 0.0, 0.0
	for i := 0; i < nBodies; i++ {
		px += bodies[i].vx * bodies[i].mass
		py += bodies[i].vy * bodies[i].mass
		pz += bodies[i].vz * bodies[i].mass
	}
	bodies[0].vx = -px / solarMass
	bodies[0].vy = -py / solarMass
	bodies[0].vz = -pz / solarMass
}

func main() {
	t0 := time.Now()

	bodies := []Body{
		// Sun
		{0, 0, 0, 0, 0, 0, solarMass},
		// Jupiter
		{4.84143144246472090e+00, -1.16032004402742839e+00, -1.03622044471123109e-01,
			1.66007664274403694e-03 * daysPerYear,
			7.69901118419740425e-03 * daysPerYear,
			-6.90460016972063023e-05 * daysPerYear,
			9.54791938424326609e-04 * solarMass},
		// Saturn
		{8.34336671824457987e+00, 4.12479856412430479e+00, -4.03523417114321381e-01,
			-2.76742510726862411e-03 * daysPerYear,
			4.99852801234917238e-03 * daysPerYear,
			2.30417297573763929e-05 * daysPerYear,
			2.85885980666130812e-04 * solarMass},
		// Uranus
		{1.28943695621391310e+01, -1.51111514016986312e+01, -2.23307578892655734e-01,
			2.96460137564761618e-03 * daysPerYear,
			2.37847173959480950e-03 * daysPerYear,
			-2.96589568540237556e-05 * daysPerYear,
			4.36624404335156298e-05 * solarMass},
		// Neptune
		{1.53796971148509165e+01, -2.59193146099879641e+01, 1.79258772950371181e-01,
			2.68067772490389322e-03 * daysPerYear,
			1.62824170038242295e-03 * daysPerYear,
			-9.51592254519715870e-05 * daysPerYear,
			5.15138902046611451e-05 * solarMass},
	}

	offsetMomentum(bodies)

	steps := 500000
	dt := 0.01
	for i := 0; i < steps; i++ {
		advance(bodies, dt)
	}

	e := energy(bodies)

	elapsed := time.Since(t0)
	fmt.Printf("energy: %.9f\n", e)
	fmt.Printf("elapsed: %.3fs\n", elapsed.Seconds())
}
